### PR TITLE
Add f16 e2e tests for vulkan

### PR DESCRIPTION
**Pull request by ThomasRaoux**

### CODE REVIEW

File: iree/test/e2e/xla_ops/BUILD [hunk 1]
```diff
@@ -125,6 +125,8 @@ iree_check_single_backend_test_suite(
             "while.mlir",
         ],
         include = ["*.mlir"],
+        # Exclude f16 tests as it is not supported by vmla.
+        exclude = ["*_f16.mlir"],
```
Comment (GMNGeoffrey) [inline iree/test/e2e/xla_ops/BUILD:129]: This kind of contradicts the big comment at the top of this file saying that everything should run on VMLA. It is still our reference backend. I'm not sure if we still generate op coverage tables from this information, but that was part of the reason for standardized naming for these files. When I needed to do something similar for integer datatypes in matmul, I created a subdirectory for "partial" support: https://github.com/google/iree/commits/main/iree/test/e2e/xla_ops/partial. Maybe we should do that here.

Comment (ThomasRaoux) [reply iree/test/e2e/xla_ops/BUILD]: Makes sense. I'm not sure if we'll want to support float16 in VMLA I'll discuss this further with Ben and the rest of the team. I moved those tests in partial for now.

File: build_tools/bazel/build_core.sh [hunk 2]
```diff
@@ -48,6 +48,8 @@ declare -a test_env_args=(
 declare -a default_build_tag_filters=("-nokokoro")
 declare -a default_test_tag_filters=("-nokokoro" "-driver=metal")
 
+# vulkan f16 tests cannot run with swiftshaders.
```
Comment (GMNGeoffrey) [inline build_tools/bazel/build_core.sh:51]:
```suggestion
# vulkan f16 tests cannot run with Swiftshader.
```

**Review by GMNGeoffrey — APPROVED** [review]

File: iree/test/e2e/xla_ops/partial/add_f16.mlir [hunk 3]
```diff
@@ -0,0 +1,7 @@
+func @tensor() attributes { iree.module.export } {
+  %0 = iree.unfoldable_constant dense<[1.0, 2.0, 3.0, 4.0]> : tensor<4xf16>
+  %1 = iree.unfoldable_constant dense<[5.0, 6.0, 7.0, 8.0]> : tensor<4xf16>
+  %result = "mhlo.add"(%0, %1) : (tensor<4xf16>, tensor<4xf16>) -> tensor<4xf16>
+  check.expect_almost_eq_const(%result, dense<[6.0, 8.0, 10.0, 12.0]> : tensor<4xf16>) : tensor<4xf16>
```
Comment (ScottTodd) [inline iree/test/e2e/xla_ops/partial/add_f16.mlir:5]: Eh.. this is tricky. We _could_ (and probably will eventually) correctly execute these tests without native support for 16 bit floating point numbers by emulating in `f32` or widening. Native support is just an optimization. I still see the point of these tests, but we should be careful with how we categorize them.

File: iree/test/e2e/xla_ops/partial/add_f16.mlir [hunk 4]
```diff
@@ -0,0 +1,7 @@
+func @tensor() attributes { iree.module.export } {
+  %0 = iree.unfoldable_constant dense<[1.0, 2.0, 3.0, 4.0]> : tensor<4xf16>
+  %1 = iree.unfoldable_constant dense<[5.0, 6.0, 7.0, 8.0]> : tensor<4xf16>
```
Comment (ScottTodd) [inline iree/test/e2e/xla_ops/partial/add_f16.mlir:3]: not sure how we handle this in other tests, but if the test is specifically exercising floating point arithmetic, shouldn't the numbers used at least have some digits after the decimal point? :P 

These tests could pass even if we truncate to `i16` :P

File: iree/test/e2e/xla_ops/partial/add_f16.mlir [hunk 5]
```diff
@@ -0,0 +1,7 @@
+func @tensor() attributes { iree.module.export } {
```
Comment (ScottTodd) [inline iree/test/e2e/xla_ops/partial/add_f16.mlir:1]: nit: also not sure how we name these in other tests, but `func @add_f16()` or `func @test_add_f16()` would be more descriptive

**Review by ScottTodd — CHANGES_REQUESTED** [review]

File: iree/test/e2e/xla_ops/partial/add_f16.mlir (hunk 3 above)
Comment (ThomasRaoux) [reply iree/test/e2e/xla_ops/partial/add_f16.mlir]: Using f32 to emulate would work but we would still need some conversion to and from float 16 for data coming from user, this is what may make the support a bit more complex. But using float32 for computation should be valid indeed.

File: iree/test/e2e/xla_ops/partial/BUILD [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    ],
+    compiler_flags = [
+        "-iree-flow-dispatch-linalg-on-tensors",
+        "-iree-codegen-spirv-experimental-linalg-on-tensors",
+        "-iree-vulkan-target-triple=valhall-g77-unknown-android10",
+    ],
+    driver = "vulkan",
+    tags = [
+        "notap",
+        "vulkan_f16",
```
[the remark this answers is not in the document]
Comment (ThomasRaoux) [reply iree/test/e2e/xla_ops/partial/BUILD]: changed the name to `vulkan_uses_vk_khr_shader_float16_int8`

File: iree/test/e2e/xla_ops/partial/add_f16.mlir (hunk 4 above)
Comment (ThomasRaoux) [reply iree/test/e2e/xla_ops/partial/add_f16.mlir]: Added some decimals :)

File: build_tools/bazel/build_core.sh [hunk 7]
```diff
@@ -48,6 +48,8 @@ declare -a test_env_args=(
 declare -a default_build_tag_filters=("-nokokoro")
 declare -a default_test_tag_filters=("-nokokoro" "-driver=metal")
 
+# SwiftShader doesn't support vk_khr_shader_float16_int8 extension.
```
Comment (ScottTodd) [inline build_tools/bazel/build_core.sh:51]:
```suggestion
# The VK_KHR_shader_float16_int8 extension is optional prior to Vulkan 1.2.
# We test on SwiftShader, which does not support this extension.
```

File: build_tools/cmake/test.sh [hunk 8]
```diff
@@ -30,6 +30,8 @@ export IREE_VULKAN_DISABLE=${IREE_VULKAN_DISABLE:-0}
 export IREE_LLVMAOT_DISABLE=${IREE_LLVMAOT_DISABLE:-0}
 # CUDA is off by default.
 export IREE_CUDA_DISABLE=${IREE_CUDA_DISABLE:-1}
+# SwiftShader doesn't support vk_khr_shader_float16_int8 extension.
```
Comment (ScottTodd) [inline build_tools/cmake/test.sh:33]:
```suggestion
# The VK_KHR_shader_float16_int8 extension is optional prior to Vulkan 1.2.
# We test on SwiftShader, which does not support this extension.
```

**Review by ScottTodd — APPROVED** [review]

File: iree/test/e2e/xla_ops/partial/add_f16.mlir (hunk 5 above)
Comment (GMNGeoffrey) [reply iree/test/e2e/xla_ops/partial/add_f16.mlir]: I think we sometimes just let the file name stand on its own and make the function names describe the specific variant of the test (like "big tensor" or whatever). With such a simple case, that makes it hard to come up with a name :-D

File: iree/test/e2e/xla_ops/partial/add_f16.mlir (hunk 3 above)
Comment (GMNGeoffrey) [reply iree/test/e2e/xla_ops/partial/add_f16.mlir]: Yeah I would hope that eventually these tests get put somewhere common to all backends. Eventually mhlo may not even be our input, so we'll probably change the way everything is organized (falls under our integration tests roadmap OKR for next quarter)

**Pull request merged**